### DISCUSSION TITLE
element.firstChild don't work in IDE but in CodeSandbox

### DISCUSSION THREAD

**Discussion by iaslfw** [vercel/next.js#48563] [Help] [answered]
### Summary

I found a problem, that I can't explain myself. Short story long: I tried to build a native draggable slider with react and positioned it in the middle of the screen. At least in my Codesandbox file. There I used this code:

```
    const windowWidth = window.innerWidth / 2;
    const itemWidth = track.firstChild.offsetWidth / 2;
    const offsetPercent = ((windowWidth - itemWidth) / window.innerWidth) * 100;

    track.style.left = `${offsetPercent}%`;
```
In codesandbox it works fine as hell. Everything like I like it to be. Now I wanted to add this test code to my real project. Build with NextJs, React & IDE VS Code.

But there I'll always get this error: "TypeError: Cannot read properties of null (reading 'offsetWidth')" (img attached below)

I can't explain myself why. The whole code is the same like in Codesandbox. 
Therefore, NextJS renders twice, 1x for server, 1x for client, I stumbled over a possible solution with optional chaining. Like so: 

  ` const itemWidth = track.firstChild?.offsetWidth / 2;`
  
But when I do it like this, it doesn't do anything. So the code doesn't run. The slider won't be positioned in the middle... 
Hope, somebody can came up with a possible solution

<img width="964" alt="Bildschirmfoto 2023-04-18 um 15 37 44" src="https://user-images.githubusercontent.com/65504843/233014405-cf796048-d43e-4734-a5ca-a9f7cca71e0c.png">

### Additional information

```js
My code is looking like this: 

useLayoutEffect(() => {
        if(!trackReference.current) return
        const track = trackReference.current
        const coordinates = coords.current
        const percent = dragPercent.current

        const windowWidth = window.innerWidth / 2;
        const itemWidth = track.firstChild.offsetWidth / 2; // <-- error here
        const offsetPercent = ((windowWidth - itemWidth) / window.innerWidth) * 100;
        const maxDistance = ((itemWidth * 2) / track.offsetWidth) * 100;

        track.style.left = `${offsetPercent}%`;
{...}
}
```

### Example

https://stackblitz.com/edit/nextjs-lpmunt?file=pages%2FSlider.jsx

**Comment by icyJoseph** [comment]
Hi,

Could you use a Stackblitz playground to create a reproduction demo? [next.new](https://next.new)?

**Reply by iaslfw** [reply]
Yessir, there you go!
https://stackblitz.com/edit/nextjs-lpmunt?file=pages%2FSlider.jsx

As a short hint. I also tried to rebuild it with {props.children}. But this doesn't work too. On page reload it still doesn't find any children and the code don't executed

**Reply by icyJoseph** [reply, marked as answer]
Hi,

Isn't the problem that you are using `next/dynamic`? Since you are using `useLayoutEffect`, the DOM for the lazy loaded parts is not yet ready. You can wrap your `ItemEl` on a `div` or something and go off that.

```jsx
  const elements = images.map((image) => {
    return (
      <div key={image.src}>
        <ItemEl src={image.src} />
      </div>
    );
  });
```

Like, if you put a `console.count` inside your `ItemEl` rendering, you'll see it counts after, `useLayoutEffect` is done.

**Reply by iaslfw** [reply]
Thanks! Mhm, I See. Interesting. Why is the problem solved all with that one div? Seem's kinda illogical 2 me🤯
